### PR TITLE
Document CI/CD security philosophy

### PR DESCRIPTION
**Pull request by malfet**
Namely:
-  when use of non-ephemeral runners is OK, vs when it is not
- Why binary build pipelines should not use distributed caching
- Why temporary CI artifacts should not be considered safe

### CODE REVIEW

**Review by seemethere — APPROVED** [review]
LGTM, happy that we finally have these in writing!

**Review by atalman — APPROVED** [review]

File: SECURITY.md [hunk 1]
```diff
@@ -61,3 +61,26 @@ If applicable, prepare your model against bad inputs and prompt injections. Some
...
+ - All binary builds are cold-start builds, i.e., distributed caching/incremental builds are not permitted. This renders builds much slower than incremental CI builds but isolates them from potential compromises of the intermediate artifacts caching systems.
+ - All upload jobs are executed in a [deployment environments](https://docs.github.com/en/actions/deployment/targeting-different-environments/using-environments-for-deployment) that are restricted to protected branches
+ - Security credentials needed to upload binaries to PyPI/conda or stable indexes `download.pytorch.org/whl` are never uploaded to repo secrets storage/environment. This requires an extra manual step to publish the release but ensures that access to those would not be compromised by deliberate/accidental leaks of secrets stored in the cloud.
+ - No binary artifacts should be published to GitHub releases pages, as these are overwritable by anyone with write permission to the repo.
```
Comment (atalman) [inline SECURITY.md:86]: We do publish [pytorch-v2.3.1.tar.gz](https://github.com/pytorch/pytorch/releases/download/v2.3.1/pytorch-v2.3.1.tar.gz) and github automatically publishes source code. 
We should probably announce for 2.4 release that we stop publishing these binaries to release page. And publish it somewhere where else instead.

Comment (malfet) [reply SECURITY.md:86]: Those are not the binaries, but rather a source code, which is probably fine, but we should continue working with GitHub to ensure fewer people can mutate the binary artifacts

**Pull request closed without merging**